src/buildstream/storage/directory.py

```diff
@@ -16,328 +17,499 @@
+    #
+    # Args:
+    #    filename: The name of the file in this directory
+    #
+    # Returns the underlying (real) file system path for the file in this
+    # directory
+    #
+    # Raises:
+    #    DirectoryError: if the backend doesn't use local files, or if
+    #                    there is no such file in this directory
+    #
+    def _get_underlying_path(self, filename) -> str:
         raise NotImplementedError()
 
-    def remove(self, *paths: str, recursive: bool = False):
-        """ Remove a file, symlink or directory. Symlinks are not followed.
-
-        Args:
-          *paths: A list of strings which are all path components.
-          recursive: True to delete non-empty directories.
-        """
+    # _get_underlying_directory()
+    #
+    # Returns the underlying (real) file system directory this
+    # object refers to.
+    #
+    # Raises:
+    #    DirectoryError: if the backend doesn't have an underlying directory
+    #
+    def _get_underlying_directory(self) -> str:
         raise NotImplementedError()
 
-    def rename(self, src: List[str], dest: List[str]):
-        """ Rename a file, symlink or directory. If destination path exists
-        already and is a file or empty directory, it will be replaced.
+    # _set_deterministic_user():
+    #
+    # Abstract method to set all files in this directory to the current user's euid/egid.
+    #
+    def _set_deterministic_user(self):
+        raise NotImplementedError()
 
-        Args:
-          *src: Source path components.
-          *dest: Destination path components.
-        """
+    # _get_size()
+    #
+    # Get an approximation of the storage space in bytes used by this directory
+    # and all files and subdirectories in it. Storage space varies by implementation
+    # and effective space used may be lower than this number due to deduplication.
+    #
+    def _get_size(self) -> int:
         raise NotImplementedError()
 
-    def _create_empty_file(self, *paths):
-        with self.open_file(*paths, mode="w"):
+    # _create_empty_file()
+    #
+    # Utility function to create an empty file
+    #
+    def _create_empty_file(self, path: str) -> None:
+        with self.open_file(path, mode="w"):
             pass
 
-
-# FileType:
-#
-# Type of file or directory entry.
-#
-class _FileType(FastEnum):
-
-    # Directory
-    DIRECTORY = 1
-
-    # Regular file
-    REGULAR_FILE = 2
-
-    # Symbolic link
-    SYMLINK = 3
-
-    # Special file (FIFO, character device, block device, or socket)
-    SPECIAL_FILE = 4
-
-    def __str__(self):
-        # https://github.com/PyCQA/pylint/issues/2062
-        return self.name.lower().replace("_", " ")  # pylint: disable=no-member
+    # _validate_path()
+    #
+    # Convenience function for backends to validate path input
+    #
+    def _validate_path(self, path: str) -> None:
+        if len(path) and path[0] == "/":
+            raise ValueError("Invalid path '{}'".format(path))
```

Comment:
This function considers empty paths valid. Do we really want this or is this necessary? Intuitively I'd expect `.` to be accepted for the current directory but the empty path to be rejected, which matches POSIX behavior.

Reply:
Ok let's do this.

Reply:
Turns out this is not particularly user friendly, not sure if we should do this.

Elements like the `import` element want to call `inputdir = inputdir.descend(self.source.strip(os.sep))`, so that if there is a "/" then the input is at the root.

Forcing the caller to specify "." here seems to introduce more special casing on the caller side than is desirable.